src/libAtomVM/intn.c

```diff
@@ -0,0 +1,631 @@
+
+/*
+ * Multiplication
+ */
+
+#ifdef USE_64BIT_MUL
+
+// Code based on Hacker's Delight book
+// Compared to the original version parameters order has been changed
+// also this version uses 64 bit multiplication
+static void mulmns32(const uint32_t u[], int m, const uint32_t v[], int n, uint32_t w[])
+{
+    uint64_t k, t, b;
+    int i, j;
+
+    for (i = 0; i < m; i++)
+        w[i] = 0;
+
+    for (j = 0; j < n; j++) {
+        k = 0;
+        for (i = 0; i < m; i++) {
+            uint64_t u_i = u[i];
+            uint64_t v_j = v[j];
+            uint64_t w_i_j = w[i + j];
+            t = u_i * v_j + w_i_j + k;
+            w[i + j] = t; // (I.e., t & 0xFFFFFFFF).
+            k = t >> 32;
+        }
+        w[j + m] = k;
+    }
+
+    // Now w[] has the unsigned product.  Correct by
+    // subtracting v*2**32m if u < 0, and
+    // subtracting u*2**32n if v < 0.
+
+    if ((int32_t) u[m - 1] < 0) {
+        b = 0; // Initialize borrow.
+        for (j = 0; j < n; j++) {
+            uint64_t w_j_m = w[j + m];
+            uint64_t v_j = v[j];
+            t = w_j_m - v_j - b;
+            w[j + m] = t;
+            b = t >> 63;
+        }
+    }
+    if ((int32_t) v[n - 1] < 0) {
+        b = 0;
+        for (i = 0; i < m; i++) {
+            uint64_t w_i_n = w[i + n];
+            uint64_t u_i = u[i];
+            t = w_i_n - u_i - b;
+            w[i + n] = t;
+            b = t >> 63;
+        }
+    }
+}
+
+void intn_mulmns(const uint32_t u[], int m, const uint32_t v[], int n, uint32_t w[])
+{
+    mulmns32(u, m, v, n, w);
+}
+
+#else
+
+// Code based on Hacker's Delight book
+// Original code with mostly no changes, except for parameters order
+static void mulmns16(const uint16_t u[], int m, const uint16_t v[], int n, uint16_t w[])
+{
+    unsigned int k, t, b;
+    int i, j;
+
+    for (i = 0; i < m; i++)
+        w[i] = 0;
+
+    for (j = 0; j < n; j++) {
+        k = 0;
+        for (i = 0; i < m; i++) {
+            t = u[i] * v[j] + w[i + j] + k;
+            w[i + j] = t; // (I.e., t & 0xFFFF).
+            k = t >> 16;
+        }
+        w[j + m] = k;
+    }
+
+    // Now w[] has the unsigned product.  Correct by
+    // subtracting v*2**16m if u < 0, and
+    // subtracting u*2**16n if v < 0.
+
+    if ((int16_t) u[m - 1] < 0) {
+        b = 0; // Initialize borrow.
+        for (j = 0; j < n; j++) {
+            t = w[j + m] - v[j] - b;
+            w[j + m] = t;
+            b = t >> 31;
+        }
+    }
+    if ((int16_t) v[n - 1] < 0) {
+        b = 0;
+        for (i = 0; i < m; i++) {
+            t = w[i + n] - u[i] - b;
+            w[i + n] = t;
+            b = t >> 31;
+        }
+    }
+    return;
+}
+
+void intn_mulmns(const uint32_t u[], int m, const uint32_t v[], int n, uint32_t w[])
+{
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+    mulmns16((const uint16_t *) u, m * 2, (const uint16_t *) v, n * 2, (uint16_t *) w);
+#elif __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+#error "Big endian not yet supported"
```

Comment:
Don't we have a big endian platform tested in CI?

Reply:
Yes, but it is not tested at the moment. That error is raised with 16 bit mul (that is useful when hardware doesn't provide 32x32 = 64 instructions) on big endian architectures.
I still need to work on that.